test/browser/components.test.js

```diff
@@ -2575,5 +2575,25 @@ describe('Components', () => {
 
 			expect(scratch.innerHTML).to.equal('<div>bar</div>');
 		});
+
+		it('should call render again when state got dirtied', () => {
+			class Persons extends Component {
+				constructor(props) {
+					super(props);
+					this.state = { x: false };
+				}
+
+				render() {
+					if (this.state.x === false) {
+						this.setState({ x: true });
```

Comment:
Curious, since the test is likely a stripped down version of the original problem: What was the scenario which lead to this PR? :slightly_smiling_face: Running effects in render seems like a known to be bad thing.

Reply:
When exiting a `render()` or functional component invocation, if state got dirtied for that component _during the execution of that function_, we can immediately re-invoke the function and use the updated result. This bypasses re-entering that update into the render queue and skips the intermediary render entirely.

```js
const Persons = (props) =>  {
  // say we need state, to allow for local mutation:
  const [checked, set] = useState(props.checked)

  const toggle = useCallback(() => {
    set(checked);
  }, [checked]);

  // when receiving a new props.checked value, update state:
  useEffect(() => {
    // here's the issue: we first render fully (to DOM) with the old value,
    // then immediately render again with the new value. Invisible but expensive.
    set(props.checked);
  }, [props.checked]);

  return <input type="checkbox" checked={checked} onClick={toggle} />
}

render(<Persons checked={false} />);
```

Reply:
@JoviDeCroock just pointed out that this actually doesn't get fixed by this PR and the suggested approach, because `useEffect()` has not yet run when we exit render.

Reply:
Also curious what scenario this is used for. At least in the above example (if theoretically this PR solved it), I typically try to avoid just copying props into local state since it duplicates the state between two components making it harder to understand who owns it and keeping each copy in sync. Though maybe it's useful for delaying value changes up the tree? perhaps for animations or debouncing?